nearbridge/contracts/NearBridge.sol

```diff
@@ -200,117 +183,137 @@ contract NearBridge is INearBridge {
         require(balanceOf[msg.sender] >= lockEthAmount, "Balance is not enough");
-        require(block.timestamp >= head.validAfter, "Wait until last block become valid");
 
         Borsh.Data memory borsh = Borsh.from(data);
         NearDecoder.LightClientBlock memory nearBlock = borsh.decodeLightClientBlock();
         require(borsh.finished(), "NearBridge: only light client block should be passed");
 
-        // 1. The height of the block is higher than the height of the current head
+        // Commit the previous block, or make sure that it is OK to replace it.
+        if (block.timestamp >= lastValidAt) {
+            if (lastValidAt != 0) {
+                commitBlock();
+            }
+        } else {
+            require(nearBlock.inner_lite.timestamp >= untrustedHead.timestamp.add(replaceDuration), "NearBridge: can only replace with a sufficiently newer block");
+        }
+
+        // Check that the new block's height is greater than the current one's.
         require(
             nearBlock.inner_lite.height > head.height,
             "NearBridge: Height of the block is not valid"
         );
 
-        // 2. The epoch of the block is equal to the epoch_id or next_epoch_id known for the current head
-        require(
-            nearBlock.inner_lite.epoch_id == head.epochId || nearBlock.inner_lite.epoch_id == head.nextEpochId,
-            "NearBridge: Epoch id of the block is not valid"
-        );
-
-        // 3. If the epoch of the block is equal to the next_epoch_id of the head, then next_bps is not None
-        if (nearBlock.inner_lite.epoch_id == head.nextEpochId) {
-            require(
-                !nearBlock.next_bps.none,
-                "NearBridge: Next next_bps should no be None"
-            );
-        }
-
-        // 4. approvals_after_next contain signatures that check out against the block producers for the epoch of the block
-        // 5. The signatures present in approvals_after_next correspond to more than 2/3 of the total stake
+        // Check that the new block is from the same epoch as the current one, or from the next one.
+        bool nearBlockIsFromNextEpoch;
         if (nearBlock.inner_lite.epoch_id == head.epochId) {
-            // The new block is from the current epoch.
-            _checkBp(nearBlock, currentBlockProducers);
+            nearBlockIsFromNextEpoch = false;
+        } else if (nearBlock.inner_lite.epoch_id == head.nextEpochId) {
+            nearBlockIsFromNextEpoch = true;
         } else {
-            // The new block is from the next epoch.
-            _checkBp(nearBlock, nextBlockProducers);
+            revert("NearBridge: Epoch id of the block is not valid");
         }
 
-        // 6. If next_bps is not none, sha256(borsh(next_bps)) corresponds to the next_bp_hash in inner_lite.
-        if (!nearBlock.next_bps.none) {
+        // Check that the new block is signed by more than 2/3 of the validators.
```

Comment:
The numbers in the comments were useful because they corresponded to the spec: https://nomicon.io/ChainSpec/LightClient.html